bitbots_ros_control/src/wolfgang_hardware_interface.cpp

```diff
@@ -8,102 +7,298 @@ namespace bitbots_ros_control {
+//todo this could be done parallel with threads for speed up
+bool WolfgangHardwareInterface::create_interfaces(ros::NodeHandle &nh,
+                                                  std::vector<std::pair<std::string, int>> dxl_devices) {
+  interfaces_ = std::vector<std::vector<hardware_interface::RobotHW *>>();
+  // init bus drivers
+  std::vector<std::string> pinged;
+  XmlRpc::XmlRpcValue port_xml;
+  nh.getParam("port_info", port_xml);
+  for (XmlRpc::XmlRpcValue::ValueStruct::const_iterator it = port_xml.begin(); it != port_xml.end(); ++it) {
+    // read bus driver specifications from config
+    XmlRpc::XmlRpcValue port_data = it->second;
+    std::string device_file = port_data["device_file"];
+    int baudrate = port_data["baudrate"];
+    int protocol_version = port_data["protocol_version"];
+    auto driver = std::make_shared<DynamixelDriver>();
+    if (!driver->init(device_file.c_str(), uint32_t(baudrate))) {
+      ROS_ERROR("Error opening serial port %s", device_file.c_str());
+      speakError(speak_pub_, "Error opening serial port");
+      sleep(1);
+      exit(1);
+    }
+    // some interface seem to produce some gitter directly after connecting. wait or it will interfere with pings
+    // sleep(1);
+    driver->setPacketHandler(protocol_version);
+    std::vector<hardware_interface::RobotHW *> interfaces_on_port;
+    // iterate over all devices and ping them to see what is connected to this bus
+    std::vector<std::tuple<int, std::string, float, float>> servos_on_port;
+    for (std::pair<std::string, int> &device : dxl_devices) {
+      std::string name = device.first;
+      int id = device.second;
+      if (std::find(pinged.begin(), pinged.end(), device.first) != pinged.end()) {
+        // we already found this and dont have to search again
+      } else {
+        ros::NodeHandle dxl_nh(nh, "device_info/" + name);
+        int model_number_specified;
+        dxl_nh.getParam("model_number", model_number_specified);
+        // some devices provide more than one type of interface, e.g. the IMU provides additionally buttons and LEDs
+        std::string interface_type;
+        dxl_nh.param<std::string>("interface_type", interface_type, "");
+        uint16_t model_number_specified_16 = uint16_t(model_number_specified);
+        uint16_t *model_number_returned_16 = new uint16_t;
+        if (driver->ping(uint8_t(id), model_number_returned_16)) {
+          // check if the specified model number matches the actual model number of the device
+          if (model_number_specified_16 != *model_number_returned_16) {
+            ROS_WARN("Model number of id %d does not match", id);
+          }
+          // ping was successful, add device correspondingly
+          // only add them if the mode is set correspondingly
+          // TODO maybe move more of the parameter stuff in the init of the modules instead of doing everything here
+          if (model_number_specified == 0xABBA && interface_type == "CORE") {
```

Comment:
No. the `model_number_specified` is already coming from the config. It is compared to this hard-coded numbers, since those are also specified hard-coded as in the workbench. It's a bit ugly, but otherwise it would be necessary to do a large rewrite of the dynamixel-workbench, which is not worth the work, in my oppinion.